File: leetcode/easy/MergeSortedArray.java
/*
Type: Easy
88. Merge Sorted Array
https://leetcode.com/problems/merge-sorted-array/?envType=study-plan-v2&envId=top-interview-150

 */

package easy;

import java.util.Arrays;

public class MergeSortedArray {
    static class Solution {
        public void merge(int[] nums1, int m, int[] nums2, int n) {
//            int[] mergedNums = new int[n+m];
//            for (int i = 0, j = 0, k = 0; ;) {
//                if (i == mergedNums.length) {
//                    break;
//                }
//                if (j < m) {
//                    mergedNums[i] = nums1[j];
//                    j++;
//                } else if (k < n) {
//                    mergedNums[i] = nums2[k];
//                    k++;
//                }
//                i++;
//            }
            for (int i = m, j = 0; i < m+n; ++i) {
                nums1[i] = nums2[j];
                j++;
            }
            Arrays.sort(nums1);
            for (int i = 0; i < nums1.length; ++i) {
                if (i == 0) {
                    System.out.print("[");
                }
                System.out.print(nums1[i]);
                if (i == nums1.length - 1) {
                    System.out.print("]");
                } else {
                    System.out.print(",");
                }
            }
        }
    }

    public static void main(String[] args) {
        int[] nums1 = new int[] {1,2,3,0,0,0};
        int m = 3;
        int[] nums2 = new int[] {2,5,6};
        int n = 3;

        Solution solution = new Solution();
        solution.merge(nums1, m, nums2, n);

    }
}

File: leetcode/top_150_interview/1_easy/merge_sorted_array.py
"""
Type: Easy
88. Merge Sorted Array
https://leetcode.com/problems/merge-sorted-array/description/?envType=study-plan-v2&envId=top-interview-150
"""
from copy import copy
from typing import List
from icecream import ic


"""
Merge Sort Solution
"""
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        i = 0
        nums1_copy = copy(nums1)
        nums2_copy = copy(nums2)
        n1 = 0
        n2 = 0
        while 1:
            if n1 == m and n2 == n:
                break
            if n1 == m:
                nums1[i] = nums2_copy[n2]
                i += 1
                n2 += 1
                continue
            if n2 == n:
                nums1[i] = nums1_copy[n1]
                i += 1
                n1 += 1
                continue
            if nums1_copy[n1] < nums2_copy[n2]:
                nums1[i] = nums1_copy[n1]
                i += 1
                n1 += 1
            else:
                nums1[i] = nums2_copy[n2]
                i += 1
                n2 += 1
        ic(nums1)


"""
Simple python Solution
"""
class Solution2:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        i = m
        for el in nums2:
            nums1[i] = el
            i += 1
        nums1.sort()
        ic(nums1)


nums1 = [1,2,3,0,0,0]
m = 3
nums2 = [2,5,6]
n = 3


# nums1 = [0]
# m = 0
# nums2 = [1]
# n = 1

solution = Solution()
ic(solution.merge(nums1, m, nums2, n))